File: 2b.py
#! /usr/bin/python3
import sys

aim   = 0
depth = 0
horiz = 0

def forward(amount):
    global horiz, depth, aim
    horiz += amount
    depth += amount * aim

def up(amount):
    global aim
    aim -= amount

def down(amount):
    global aim
    aim += amount

with open(sys.argv[1], 'r') as fh:
    for line in fh:
        (direction, amount) = line.split()
        {
            'forward': forward,
            'up'     : up,
            'down'   : down
        }[direction](int(amount))
print(depth * horiz)

File: 2b.pl
#!/usr/bin/perl
use warnings;
use strict;
use feature qw{ say };

use ARGV::OrDATA;

my ($aim, $depth, $horiz) = (0, 0, 0);
while (<>) {
    my ($direction, $amount) = split;
    {
        forward => sub { $horiz += $amount; $depth += $amount * $aim },
        up      => sub { $aim -= $amount },
        down    => sub { $aim += $amount }
    }->{$direction}();
}
say $depth * $horiz;

__DATA__
forward 5
down 5
forward 8
up 3
down 8
forward 2
